pkg/controllers/dnspolicy/dns_helper.go

```diff
@@ -36,44 +31,57 @@ var (
 	ErrAlreadyAssigned      = fmt.Errorf("managed host already assigned")
 )
 
-type Service struct {
-	controlClient client.Client
+type dnsHelper struct {
+	client.Client
 }
 
-func NewService(controlClient client.Client) *Service {
-	return &Service{controlClient: controlClient}
+// GetManagedZoneForHost returns a ManagedZone and subDomain for the given host if one exists.
+//
+// Currently, this returns the first matching ManagedZone found in the traffic resources own namespace
+func (dh *dnsHelper) getManagedZoneForHost(ctx context.Context, host string, t traffic.Interface) (*v1alpha1.ManagedZone, string, error) {
+	var managedZones v1alpha1.ManagedZoneList
+	if err := dh.List(ctx, &managedZones, client.InNamespace(t.GetNamespace())); err != nil {
+		log.FromContext(ctx).Error(err, "unable to list managed zones in traffic resource NS")
+		return nil, "", err
+	}
+	return findMatchingManagedZone(host, host, managedZones.Items)
 }
 
-func (s *Service) GetManagedHosts(ctx context.Context, traffic traffic.Interface) ([]v1alpha1.ManagedHost, error) {
-	var managed []v1alpha1.ManagedHost
-	for _, host := range traffic.GetHosts() {
-		managedZone, subDomain, err := s.GetManagedZoneForHost(ctx, host, traffic)
-		if err != nil && !errors.Is(err, ErrNoManagedZoneForHost) {
-			return nil, err
-		}
-		if managedZone == nil {
-			// its ok for no managedzone to be present as this could be a CNAME or externally managed host
-			continue
-		}
-		dnsRecord, err := s.GetDNSRecord(ctx, subDomain, managedZone, traffic)
-		if err != nil && !k8serrors.IsNotFound(err) {
-			return nil, err
-		}
-		managedHost := v1alpha1.ManagedHost{
-			Host:        host,
-			Subdomain:   subDomain,
-			ManagedZone: managedZone,
-			DnsRecord:   dnsRecord,
-		}
+func findMatchingManagedZone(originalHost, host string, zones []v1alpha1.ManagedZone) (*v1alpha1.ManagedZone, string, error) {
+	if len(zones) == 0 {
+		return nil, "", fmt.Errorf("%w : %s", ErrNoManagedZoneForHost, host)
+	}
+	host = strings.ToLower(host)
+	//get the TLD from this host
+	tld, _ := publicsuffix.PublicSuffix(host)
 
-		managed = append(managed, managedHost)
+	//The host is just the TLD, or the detected TLD is not an ICANN TLD
+	if host == tld {
+		return nil, "", fmt.Errorf("no valid zone found for host: %v host is tld %v", originalHost, tld)
 	}
-	return managed, nil
+
+	hostParts := strings.SplitN(host, ".", 2)
+	if len(hostParts) < 2 {
+		return nil, "", fmt.Errorf("no valid zone found for host: %s", originalHost)
+	}
+
+	zone, ok := slice.Find(zones, func(zone v1alpha1.ManagedZone) bool {
+		return strings.ToLower(zone.Spec.DomainName) == host
+	})
+
+	if ok {
+		subdomain := strings.Replace(strings.ToLower(originalHost), "."+strings.ToLower(zone.Spec.DomainName), "", 1)
+		return &zone, subdomain, nil
+	} else {
+		parentDomain := hostParts[1]
+		return findMatchingManagedZone(originalHost, parentDomain, zones)
+	}
+
```

Comment:
There are changes made to this on main that aren't reflected here https://github.com/Kuadrant/multicluster-gateway-controller/pull/350